src/main/java/gov/ornl/ssm/ml/ui/views/TrainingView.java

```diff
@@ -85,82 +100,102 @@ public void init() {
-				
-				//Read each abbreviated model
-				for(Model temp : tempModels) {
 					
-					//Count the new model
-					seen += 1;
+					// If the token is expired, force a spring-boot token refresh by refreshing the entire page.
+		            if(client.getAccessToken().getExpiresAt().isBefore(Instant.now())) {
+		            	SecurityContextHolder.getContext().setAuthentication(null);
+		            	UI.getCurrent().navigate("training");
+		            }
 					
-					String modelString = "";
+					HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+					conn.setRequestProperty("Authorization", "Bearer " + client.getAccessToken().getTokenValue());
+					conn.setRequestMethod("GET");
+					BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
 					
-					//Change the urls to point to the fuseki host
-					String urlString = temp.getUrl();
-					urlString = config.getFusekiHost() + urlString.substring(urlString.indexOf(".gov") + 4);
-					if(urlString.endsWith("/")) {
-						urlString = urlString.substring(0, urlString.lastIndexOf("/"));
-					}
+					jsonModels = "";
 					
-					// Get the model digest from the backend
-					URL modelUrl = new URL(urlString);
-					HttpURLConnection modelConn = (HttpURLConnection) modelUrl.openConnection();
-					modelConn.setRequestMethod("GET");
-					reader = new BufferedReader(new InputStreamReader(modelConn.getInputStream()));
-
 					// Read the results into the string
-					line = reader.readLine();
-
+					String line = reader.readLine();
+	
 					while (line != null) {
-						modelString += line;
+						jsonModels += line;
 						line = reader.readLine();
 					}
 					
-				
-					models.add(mapper.readValue(modelString, Model.class));
-				}
-				
+	
+					// Get the total number of models to download
+					JsonNode digest = mapper.readTree(jsonModels);
+					total = digest.get("total").asInt();
+					
+					// Get the first result page's url
+					String next = digest.get("next").asText();
+					url = new URL(next);
+					
+					// Strip out the pagination data, leaving only the raw models
+					String modelsString = mapper.writeValueAsString(digest.get("data"));
+	
+					// Convert the model data back to JSON, then read the JSON into classes
+					List<Model> tempModels = mapper.readValue(modelsString,
+							mapper.getTypeFactory().constructCollectionType(List.class, Model.class));
+					
+					//Read each abbreviated model
+					for(Model temp : tempModels) {
+						
+						//Count the new model
+						seen += 1;
+						
+						String modelString = "";
+						
+						//Change the urls to point to the fuseki host
+						String urlString = temp.getUrl();
+						if(urlString.contains(".gov")) {
+							urlString = config.getFusekiHost() + urlString.substring(urlString.indexOf(".gov") + 4);
+						}
+						if(urlString.endsWith("/")) {
+							urlString = urlString.substring(0, urlString.lastIndexOf("/"));
+						}
+						
```

Comment:
So is this pulling out the model URLs and changing to the Fuseki hostname to grab the models (instead of the main URL hostname like `ssm-dev.ornl.gov`)?

Reply:
Probably a better way we can do this but I have no idea.

This is perfectly fine for me now.

Just wondering how others usually get over this. :thinking: 

Thanks

Reply:
The correct way is to make sure that the machine can reach its own proxied web address, but that VM configuration issue isn't a high priority at the moment.